pprl/layers/layer_wrappers.py

```diff
@@ -0,0 +1,546 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Wrappers for fluid.layers so that the layers can share parameters conveniently.
+"""
+
+from paddle.fluid.layers import *
+from paddle.fluid.param_attr import ParamAttr
+import paddle.fluid.layers as layers
+import warnings
+import inspect
+
+all_wrapped_layers = [
+    "create_parameters", "fc", "embedding", "dynamic_lstm", "dynamic_lstmp",
+    "dynamic_gru", "sequence_conv", "conv2d", "conv2d_transpose", "lstm_unit",
+    "row_conv"
+]
+
+
+class LayerCounter:
+    custom = {}
+    create_parameter = 0
+    fc = 0
+    embedding = 0
+    dynamic_lstm = 0
+    dynamic_lstmp = 0
+    dynamic_gru = 0
+    sequence_conv = 0
+    conv2d = 0
+    conv2d_transpose = 0
+    lstm_unit = 0
+    row_conv = 0
+
+
+class LayerFunc(object):
+    def __init__(self, param_attr=False, bias_attr=False):
+        self.param_attr = param_attr
+        self.bias_attr = bias_attr
+
+    @property
+    def param_name(self):
+        if self.param_attr:
+            return self.param_attr.name
+        else:
+            return None
+
+    @property
+    def bias_name(self):
+        if self.bias_attr:
+            return self.bias_attr.name
+        else:
+            return None
+
+    def parameters(self):
+        return (self.param_attr, self.bias_attr)
+
+
+def get_set_paras(set_paras):
+    param_name, bias_name = None, None
+    if set_paras is not None:
+        assert (type(set_paras) is tuple) and len(set_paras) == 2
+        param_attr, bias_attr = set_paras
+        if param_attr:
+            param_name = param_attr.name
+        if bias_attr:
+            bias_name = bias_attr.name
+    return param_name, bias_name
+
+
+def check_or_replace_name(name, new_name, attr):
+    name = (new_name if name is None else name)
+    ## if this para is not used
+    if attr == False:
+        return False
+
+    if attr is None:
+        return ParamAttr(name=name)
+
+    assert attr.name is None, \
+        "Do not set parameter name for pprl.layers; leave it as None"
+    attr.name = name
+    return attr
+
+
+def update_attr_name(name, param_name, field_name, attr, suffix,
+                     counter_increase):
+    if name is None:
+        name = field_name
+    else:
+        name = "_" + name
+        field_name = "custom"
+
+    if field_name == "custom":
+        custom_counter = getattr(LayerCounter, field_name)
+        if not name in custom_counter:
+            custom_counter[name] = 0
+        idx = custom_counter[name]
+        if counter_increase:
+            custom_counter[name] += 1
+        setattr(LayerCounter, field_name, custom_counter)
+    else:
+        idx = getattr(LayerCounter, field_name)
+        if counter_increase:
+            setattr(LayerCounter, field_name, idx + 1)
+
+    new_name = "%s_%d%s" % (name, idx, suffix)
```

Comment:
And It's might be better to use fluid.unique_name.generate() to generate the name to ensure unique name.